x/batching/keeper/abci.go

```diff
@@ -0,0 +1,85 @@
+package keeper
+
+import (
+	"encoding/binary"
+	"encoding/hex"
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/sedaprotocol/seda-chain/x/batching/types"
+
+	"github.com/cometbft/cometbft/crypto/merkle"
+)
+
+func (k Keeper) EndBlock(ctx sdk.Context) (err error) {
+	// Use defer to prevent returning an error, which would cause
+	// the chain to halt.
+	defer func() {
+		// Handle a panic.
+		if r := recover(); r != nil {
+			k.Logger(ctx).Error("recovered from panic in batching EndBlock", "err", r)
+		}
+		// Handle an error.
+		if err != nil {
+			k.Logger(ctx).Error("error in batching EndBlock", "err", err)
+		}
+		err = nil
+	}()
+
+	batch, err := k.ConstructBatch(ctx)
+	if err != nil {
+		panic(err)
+	}
+	fmt.Println(batch)
+
+	err = k.SetBatch(ctx, batch)
+	if err != nil {
+		panic(err)
+	}
+
+	return nil
+}
+
+func (k Keeper) ConstructBatch(ctx sdk.Context) (types.Batch, error) {
+	curBatchNum, err := k.GetCurrentBatchNum(ctx)
+	if err != nil {
+		return types.Batch{}, err
+	}
+
+	// Construct data result tree.
+	// Fetch batch-ready data requests.
+	// TODO: Deal with offset and limits. (#313)
+	// queryRes, err := k.wasmViewKeeper.QuerySmart(ctx, coreContract, []byte(`{"get_data_requests_by_status":{"status": "tallied", "offset": 0, "limit": 100}}`))
+	// if err != nil {
+	// 	return err
+	// }
+	// if string(queryRes) == "[]" {
+	// 	return nil
+	// }
+
+	// Leaves are SHA-256 hashes of JSON-serialized data result structs.
```

Comment:
Aren't the data result IDs already hashes of the data result?

Reply:
We confirmed that they are, so we will use the data result ID as the tree's leaves.